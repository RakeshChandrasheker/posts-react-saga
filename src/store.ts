import { createStore, applyMiddleware } from 'redux';
import createSagaMiddleware from 'redux-saga';
import { composeWithDevTools } from 'redux-devtools-extension';
import rootSaga from './store/saga';
import reducers from './store/reducers';


const sagaMiddleware = createSagaMiddleware();

const store = createStore(
  reducers,
  composeWithDevTools(applyMiddleware(sagaMiddleware))
);

sagaMiddleware.run(rootSaga);

export default store;