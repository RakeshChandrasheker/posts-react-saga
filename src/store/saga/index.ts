import { PostsActionTypes } from '../action-types/index';
import { put, takeEvery, all, call } from 'redux-saga/effects';
import { downloadPostsErrorAction, downloadPostsOkAction, editPostErrorAction, editPostOkAction } from '../actions/posts-action';
import { editPostDB, retrievePostsDB } from '../api-calls';
import { displayToastMessage } from '../../components/notification/Notification';

// Retrieve posts
// worker saga
function* retrievePosts() {
  try {
    const { data } = yield call(retrievePostsDB);
    yield downloadPostsOkAction(data);
  } catch (error) {
    yield put(downloadPostsErrorAction());
    displayToastMessage('error', 'An error ocurred')
  }
}

// watcher saga
function* retrievePostsSaga() {
  yield takeEvery(PostsActionTypes.BEGIN_POSTS_DOWNLOAD, retrievePosts);
}

// Edit post
// worker saga
function* editPost(action: any) {
  const { post } = action;
  try {
    yield call(editPostDB, post);
    yield editPostOkAction(post);
    displayToastMessage('success', 'Post edited successfully!')
  } catch (error) {
    yield editPostErrorAction();
    displayToastMessage('error', 'An error ocurred')
  }
}

// watcher saga
function* editPostSaga() {
  yield takeEvery(PostsActionTypes.BEGIN_EDIT_POST, editPost);
}

export default function* rootSaga() {
  yield all([
    retrievePostsSaga(),
    editPostSaga()
  ]);
}
