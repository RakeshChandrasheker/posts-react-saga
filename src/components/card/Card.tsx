import { FunctionComponent, useState } from "react"
import { IPost } from "../../store/state-types"
import Pencil from '../../assets/icon/pencil.png';
import './Card.scss';
import BasicModal from "../modal/Modal";
import { retrievePostEditAction } from "../../store/actions/posts-action";

const Card: FunctionComponent<IPost> = (props): JSX.Element => {
  const [isModalOpen, setModal] = useState<boolean>(false)
  const { title, description } = props

  const openModal = (): void => {
    retrievePostEditAction(props);
    setModal(true)
  };

  const closeModal = (): void => {
    setModal(false)
  };

  const saveChanges = (): void => {
    setModal(false)
  };

  const modal = (): JSX.Element => {
    return (
      <BasicModal
        closeModal={closeModal}
        isModalOpen={isModalOpen}
        saveChanges={saveChanges}
      />
    );
  };

  return (
    <div className="card">
      <div className='titleWithIcon'>
        <div className='title'>{title}</div>
        <img className='pencilIcon' onClick={openModal} src={Pencil} alt="Pencil icon" />
      </div>
      <div className='description'>{description}</div>
      {isModalOpen && modal()}
    </div>
  )
}

export default Card
