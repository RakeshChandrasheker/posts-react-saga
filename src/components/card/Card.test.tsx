import React from 'react';
import { shallow } from 'enzyme';
import { IPost } from '../../store/state-types';
import Card from './Card';

describe('<Card />', () => {
  const setState = jest.fn();
  const useStateMock: any = (initState: any) => [initState = true, setState];
  jest.spyOn(React, 'useState').mockImplementation(useStateMock);

  const cardProps: IPost = {
    title: 'Mock_Title',
    description: 'Mock_Description'
  }
  const wrapper = shallow(<Card {...cardProps} />);


  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should trigger modal', () => {
    const pencilIcon = wrapper.find('.pencilIcon');
    pencilIcon.simulate('click');
    expect(setState).toHaveBeenCalledTimes(1);
  });


  it('should Match Snapshot', () => {
    expect(wrapper.getElement()).toMatchSnapshot();
  });

  afterAll(() => {
    wrapper.unmount();
  });
});
