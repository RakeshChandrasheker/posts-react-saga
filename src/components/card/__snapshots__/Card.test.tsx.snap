// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> should Match Snapshot 1`] = `
<div
  className="card"
>
  <div
    className="titleWithIcon"
  >
    <div
      className="title"
    >
      Mock_Title
    </div>
    <img
      alt="Pencil icon"
      className="pencilIcon"
      onClick={[Function]}
      src="pencil.png"
    />
  </div>
  <div
    className="description"
  >
    Mock_Description
  </div>
  <BasicModal
    closeModal={[Function]}
    isModalOpen={true}
    saveChanges={[Function]}
  />
</div>
`;
